onnxruntime/core/providers/xnnpack/xnnpack_provider_factory.cc

```diff
@@ -0,0 +1,30 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+#include "core/providers/xnnpack/xnnpack_provider_factory.h"
+
+#include "core/providers/xnnpack/xnnpack_execution_provider.h"
+#include "core/session/abi_session_options_impl.h"
+
+namespace onnxruntime {
+
+struct XnnpackProviderFactory : IExecutionProviderFactory {
+  XnnpackProviderFactory() = default;
+  std::unique_ptr<IExecutionProvider> CreateProvider() override;
+};
+
+std::unique_ptr<IExecutionProvider> XnnpackProviderFactory::CreateProvider() {
+  XnnpackExecutionProviderInfo info{true};
+  return std::make_unique<XnnpackExecutionProvider>(info);
+}
+
+std::shared_ptr<IExecutionProviderFactory> CreateExecutionProviderFactory_Xnnpack() {
+  return std::make_shared<XnnpackProviderFactory>();
+}
+
+}  // namespace onnxruntime
+
+ORT_API_STATUS_IMPL(OrtSessionOptionsAppendExecutionProvider_Xnnpack, _In_ OrtSessionOptions* options) {
+  options->provider_factories.push_back(onnxruntime::CreateExecutionProviderFactory_Xnnpack());
```

Comment:
consider wrapping in API_IMPL_BEGIN/API_IMPL_END just to be safe in case there are exceptions

Reply:
This is a static publicly exported function rather than part of OrtApis so the begin/end pair doesn't work for it as it returns void not OrtStatus.

Will add the ability to enable via OrtApis in a separate PR.

Reply:
it looks like it returns an OrtStatusPtr
https://github.com/microsoft/onnxruntime/blob/1f2c92673b2edf75460478d406f84aead7a140f3/include/onnxruntime/core/session/onnxruntime_c_api.h#L131-L133
don't need to change in this PR

Reply:
Will address in the PR that adds enabling xnnpack to the various language bindings. 

There is some structural weirdness with API_IMPL_BEGIN/END. It's defined in framework/error_code_helper.h but includes the C API from session (framework is a lower level library than session so this is somewhat backwards), and API_IMPL_END uses OrtApis::CreateStatus which requires including session/ort_apis.h. It was the lack of ort_apis.h that caused an error when I initially tried to add the macro usage).

I assume this is to keep the C API header nice and clean.